src/crimson/thread/ThreadPool.h

```diff
@@ -0,0 +1,93 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*- 
+// vim: ts=8 sw=2 smarttab
+
+#include <atomic>
+#include <condition_variable>
+#include <tuple>
+#include <type_traits>
+#include <boost/lockfree/queue.hpp>
+#include <boost/optional.hpp>
+#include <core/future.hh>
+#include <core/semaphore.hh>
+
+#include "Condition.h"
+
+namespace ceph::thread {
+
+struct WorkItem {
+  virtual ~WorkItem() {}
+  virtual void process() = 0;
+};
+
+template<typename Func, typename T = std::invoke_result_t<Func>>
+struct Task final : WorkItem {
+  Func func;
+  boost::optional<T> result;
+  ceph::thread::Condition on_done;
+  seastar::semaphore& free_slot;
+public:
+  Task(seastar::semaphore& fs, Func&& f)
+    : func(std::move(f)),
+      free_slot(fs)
+  {}
+  void process() override {
+    result = func();
+    on_done.notify();
+  }
+  seastar::future<T> get_future() {
+    return on_done.wait().then([this] {
+        free_slot.signal();
```

Comment:
strictly speaking, we can signal the fiber waiting for `free_slot` once the task is popped from the queue, but it'd be easier for us to signal a *seastar* thread using a single instance of sync primitive. an FD does not work here, as it is removed from epoll watch set once it is signaled, but we could have multiple fibers waiting for `free_slot`.

so, the downside of this solution is that we parallel level is decremented by one. the advantage of it is, "it works" (r).